plugin/src/compilation/decompiler.ml

```diff
@@ -156,36 +154,102 @@ let rec simpl (tacs : tact list) : tact list =
+  (* this is bugged, see intros. vs intro x. *)
+  match tacs with
+  | Intros [] :: tacs' -> intros_induction tacs'
+  | Intros xs :: Induction (env, x, ns, prefix, goals) :: tacs' ->
+     let prefix' = intros_induction prefix in
+     let goals' = List.map intros_induction goals in
+     let tacs'' = intros_induction tacs' in
+     let rest = Induction (env, x, ns, prefix', goals') :: tacs'' in
+     Option.default (Intros xs :: rest)
+       (guard (xs <> []) >>= fun _ ->
+        let last = List.hd (List.rev xs) in
+        try_rel x >>= fun idx ->
+        let n = expect_name (fst (rel_name_type (lookup_rel idx env))) in
+        guard (Id.equal last n) >>= fun _ ->
+        Some rest)
+  | Split (prefix, goals) :: tacs' ->
+     Split (intros_induction prefix, List.map intros_induction goals) :: intros_induction tacs'
+  | Induction (env, x, ns, prefix, goals) :: tacs' ->
+     Induction (env, x, ns, intros_induction prefix, List.map intros_induction goals)
+     :: intros_induction tacs'
+  | tac :: tacs' -> tac :: intros_induction tacs'
+  | [] -> []   
+
+(* Returns true if the given tactic solves the type
+   of the given term. *)
+let solves env sigma tac trm =
+  try
+    let goal = (Typeops.infer env trm).uj_type in
+    let p = Proof.start sigma [(env, EConstr.of_constr goal)] in
+    let (p', _) = Proof.run_tactic (Global.env()) tac p in
+    let (result, _, _, _, _) = Proof.proof p' in
+    List.length result == 0
+  with _ ->
+    false
+
+(* Given the list of tactics and their corresponding string
+   expressions, try to solve the given goal (type of trm),
+   return None otherwise. *)
+let rec try_solve env sigma opts trm =
+  match opts with
+  | [] -> None
+  | (tac, expr) :: opts' ->
+     if solves env sigma tac trm
+     then Some (Expr expr)
+     else try_solve env sigma opts' trm
+        
+(* Performs the bulk of decompilation on a proof term.
    Returns a list of tactics. *)
-let rec first_pass env sigma trm =
-  (* Apply single beta reduction to terms that *might*
+let rec first_pass env sigma opts trm =
+  (* Apply single reduction to terms that *might*
        be in eta expanded form. *)
   let trm = Reduction.whd_betaiota env trm in
-  let choose f x = Option.default [Apply (env, trm)] (f x (env, sigma)) in
-  match kind trm with
-  (* "fun x => ..." -> "intro x." *)
-  | Lambda (n, t, b) ->
-     let name = fresh_name env n in
-     let env = push_local (Name name, t) env in
-     Intro name :: first_pass env sigma b
-  (* Match on well-known functions used in the proof. *)
-  | App (f, args) ->
-     choose (rewrite <|> induction <|> left <|> right <|> split
-             <|> reflexivity <|> symmetry <|> exists) (f, args)
-  (* Hypothesis transformations or generation tactics. *)
-  | LetIn (n, valu, typ, body) ->
-     choose (rewrite_in <|> apply_in <|> pose) (n, valu, typ, body)
-  (* Remainder of body, simply apply it. *)
-  | _ -> [ Apply (env, trm) ]
-  
+  let solved = try_solve env sigma opts trm in
```

Comment:
I really think it makes more sense to move this into the second pass as iterative replacement, rather than try it in the first pass.

Reply:
If you think it's better in the first past, though, feel free to justify it---I'm flexible on this point, since you're the one writing the code :)